packages/flutter/lib/src/cupertino/switch.dart

```diff
@@ -507,12 +502,17 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
     );
     final double thumbCenterY = offset.dy + size.height / 2.0;
 
+    canvas.save();
```

Comment:
Most of the time we just use `context.pushClipRRect`. An example:
https://github.com/flutter/flutter/blob/ab26e1b4a7cdefb5b9b2ba53e35c68ec3116f0e3/packages/flutter/lib/src/rendering/proxy_box.dart#L1371

Relaying @liyuqian 's answer, if it is just the shadow you want to clip off, it's probably not worth doing `saveLayer` (as there's no overlapping draw calls at the edge of the `RRect`, if I understand it correctly).